File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.js
const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n")
  .map((el) => el.split(" ").map(Number));

const [tempN, haveCards, tempM, getCards] = input;

const N = +tempN.join("");
haveCards.sort((a, b) => a - b);

let answer = [];

for (let card of getCards) {
  let upperBound = upperBoundSearch(haveCards, card, 0, N);
  let lowerBound = lowerBoundSearch(haveCards, card, 0, N);

  answer.push(upperBound - lowerBound);
}

console.log(...answer);

function lowerBoundSearch(list, target, left, right) {
  let mid = 0;

  while (left < right) {
    mid = Math.floor((left + right) / 2);

    if (list[mid] >= target) right = mid;
    else {
      left = mid + 1;
    }
  }

  return right;
}

function upperBoundSearch(list, target, left, right) {
  let mid = 0;

  while (left < right) {
    mid = Math.floor((left + right) / 2);

    if (list[mid] > target) {
      right = mid;
    } else left = mid + 1;
  }

  return right;
}

File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.java
import java.io.*;
import java.util.*;

class Main {
    
    static BufferedReader br;
    static StringBuilder sb;
    
    static int N; // 가지고 있는 숫자 카드 수
    static int[] cards;
    
    static int M;
    static int[] searchList;
    
	public static void main(String[] args) throws Exception {
        
		br = new BufferedReader(new InputStreamReader(System.in));
		sb = new StringBuilder();
        
        N = Integer.parseInt(br.readLine());
        
        cards = new int[N];
        String[] input = br.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            cards[i] = Integer.parseInt(input[i]);
        }
        
        M = Integer.parseInt(br.readLine());
        
        searchList = new int[M];
        input = br.readLine().split(" ");
        for (int i = 0; i < M; i++) {
            searchList[i] = Integer.parseInt(input[i]);
        }
        
        // --------------------------- input ---------------------------
        
        Arrays.sort(cards);
        
        for (int target: searchList) {
            int count = upperBound(target) - lowerBound(target);
            
            sb.append(count).append(" ");
        }
        
        System.out.println(sb);
        
	}
    
    static int lowerBound(int target) {
        
        int left = 0;
        int right = N;
        
        while (left < right) {
            int mid = (left + right) / 2;
            
            if (cards[mid] < target) left = mid + 1;
            else right = mid;
        }
        
        return left;
        
    }
    
    static int upperBound(int target) {
        
        int left = 0;
        int right = N;
        
        while (left < right) {
            int mid = (left + right) / 2;
            
            if (cards[mid] <= target) left = mid + 1;
            else right = mid;
        }
        
        return left;
        
    }
    
}